### DISCUSSION TITLE
Formatting Prisma 2 schema programmatically in JavaScript

### DISCUSSION THREAD

**Discussion by jhaemin** [prisma/prisma#2834] [Q&A] [answered]
Is there any way to programmatically format Prisma 2 schema in JavaScript code?

Like

```js
import { format } from 'prisma'

const schema = ``

const formattedSchema = format(schema)
```

**Comment by ryands17** [comment, marked as answer]
Hey @jhaemin :wave: 
Unfortunately this is not supported in Prisma currently. Could you create a Feature Request [here](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=feature_request.md&title=) describing the above use case.

**Reply by Jolg42** [reply]
You can actually use it today 😃
```
const { formatSchema } = require('@prisma/sdk');

// and then use it like
const output = await formatSchema({ schemaPath })
```

Under the hood this will call the prisma-fmt binary and return the output.

**Comment by timleslie** [comment]
I was able to achieve this with:

```
const execa = require('execa');
const { resolveBinary } = require('@prisma/sdk');

const prismaSchema = (
  await execa(await resolveBinary('prisma-fmt'), ['format'], {
    env: { ...process.env, RUST_BACKTRACE: '1' },
    maxBuffer: 1000 * 1000 * 1000,
    input: `unformatted schema goes here`
  })
).stdout;
```

It's not ideal, but it gets the job done for now.

Feature request submitted: https://github.com/prisma/prisma/issues/3265

**Reply by Jolg42** [reply]
Can be done like this 

```
const { formatSchema } = require('@prisma/sdk');

// and then use it like
const output = await formatSchema({ schemaPath })
```

Under the hood this will call the prisma-fmt binary and return the output.

**Reply by timleslie** [reply]
This requires having the file written out to the file-system. Ideally I'd like to be able to do all this in memory before having to write to disk. In my use case I'm generating a collection of schemas and want to be able to take a hash of the (formatted) schema and look it up and only write to disk if it's a new schema. Writing to disk just to format the file so that I can check whether I need to write to disk is a bit of overkill. (I also realise that what I'm doing is perhaps outside of what could be considered regular usage of the library :-) )

**Reply by Jolg42** [reply]
Ok I understand your point, we'll see what we can do in the feature request issue!

**Reply by Jolg42** [reply]
I think it's a valid use-case!

So since dev version `2.6.0-dev.16` you can now use it like:

```
await formatSchema({ 
	schema: `
		// My Prisma Schema
	`
})
```

or with a path
```
await formatSchema({ schemaPath: '..../schema.prisma' })
```

It will be part of the 2.6.0 release.

**Reply by timleslie** [reply]
This new updated API works perfectly, thanks @Jolg42 🚀

**Comment by ludralph** [comment]
Hi @jhaemin,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
